awx/main/scheduler/task_manager_instances.py

```diff
@@ -0,0 +1,236 @@
+# Copyright (c) 2015 Ansible, Inc.
+# All Rights Reserved
+
+# Python
+import logging
+import heapq
+from types import SimpleNamespace
+
+# Django
+from django.conf import settings
+
+# AWX
+from awx.main.scheduler.dependency_graph import DependencyGraph
+from awx.main.managers import InstanceGroupManager
+from awx.main.models import (
+    Instance,
+    InstanceGroup,
+)
+
+
+logger = logging.getLogger('awx.main.scheduler')
+
+
+class PrioritizableNode:
+    def __init__(self, remaining_capacity, hostname):
+        self.remaining_capacity = remaining_capacity
+        self.hostname = hostname
+
+    def __gt__(self, other):
+        return self.remaining_capacity > other.remaining_capacity
+
+    def __lt__(self, other):
+        return self.remaining_capacity < other.remaining_capacity
+
+    def __eq__(self, other):
+        if hasattr(other, 'hostname'):
+            other = other.hostname
+        return self.hostname == other
+
+    def __hash__(self):
+        return hash(self.hostname)
+
+    def __str__(self):
+        return f"{self.hostname}: {self.remaining_capacity} remaining capacity"
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class PrioritizedNodes:
+    def __init__(self):
+        self.nodes = set()
+
+    def add(self, node):
+        self.nodes.add(node)
+
+    def __getitem__(self, hostname):
+        if hostname not in self.nodes:
+            raise KeyError(f"No control node with hostname {hostname} found")
+        for node in self.nodes:
+            if node == hostname:
+                return node
+
+    def update(self, node):
+        """The __hash__ function for PrioritizableNodes just compares hostnames.
+
+        We can add/remove the node to update object in PrioritizedNodes to have right capacity.
+        """
+        self.nodes.remove(node)
+        self.nodes.add(node)
+
+    def best_node(self):
+        heap = list(self.nodes)
+        heapq.heapify(heap)
+        return heapq.nlargest(1, heap).pop()
```

Comment:
I would really prefer to keep the paradigms we have of

1. try to find a node where the job "fits"
2. find largest idle instance

I don't see why that won't work for control nodes as well. You don't need to reuse the existing logic for this, you could write new methods that do this, but the selection algorithm is something I would prefer to not alter in this refacoring.

Reply:
Is there a technical reason for your preference?